wicket/src/defaults/dimensions.rs

```diff
@@ -4,4 +4,51 @@
 
 //! Default dimensions for various widgets common across screens
 
+use tui::layout::Rect;
+
 pub const MENUBAR_HEIGHT: u16 = 5;
+
+// Extension trait for Rects that provide centering and placement capabilities
+pub trait RectExt {
```

Comment:
I didn't understand what these methods did without looking at their implementations. A couple specific things that confused me:

1. The names sound like they'd take `&mut self`, but all of these leave `self` unmodified and return a new `Rect`.
2. In the doc comments that refer to a `container`, it's actually `self` that is the container, right?

Taking a stab at addressing both of these, maybe something like

```rust
/// Create a new `Rect` with the given width centered horizontally within `self`.
/// Panics if `width > self.width`.
fn centered_horiz(self, width: u16) -> Self;

/// Create a new `Rect` with the given height centered vertically within `self`.
/// Panics if `height > self.height`.
fn centered_vert(self, height: u16) -> Self;

/// Derive a new rectangle that begins at the given `x` position, but has the same
/// ending `x` position and vertical position/dimension as `self`. Panics if `x > self.width`.
fn ???(self, left: x: u16) -> Self;

/// Derive a new rectangle that begins at the given `y` position, but has the same
/// ending `y` position and horizontal position/dimension as `self`. Panics if `y > self.height`.
fn ???(self, bottom: y: u16) -> Self;
```

The latter two methods seem difficult to name. Maybe something like `shrink_with_left()` / `shrink_with_bottom()`?

Reply:
Yeah....

Originally, I called this trait `Center` and only had the first two methods named `centered_horizontally` and `centered_vertically` and they did indeed take `&mut self`. I shortened the names with x and y, and took self so I could assign to a variable directly without having to declare it mutably on a separate line. I also realized that in deduping some common Rect behavior for centering I could also dedup some other functions, and this could become a more general extension trait. Unfortunately, as you noticed it's not exactly straightforward to figure out what to call those things. 

As you probably surmised, when positioning a rectangle by moving it down or to the right, you have to resize it negatively with an equal amount so that it doesn't exceed the size of the terminal. That causes a panic. Since, we are always combining those two operations and leaving the second one off always causes a panic I tried to wrap them here. My goal was to indicate that I was moving this rectangle on the screen to a given location. But it does also shrink it as required. I think we should indicate this is "placing" the rectangle  and shrinking it to fit its container, which is assumed to always be correctly sized before the move. Something like `place_left_and_shrink_to_fit` is more explicit, but I still feel like the reader would end up looking to see what the hell the call did. With that in mind, I'm not sure we can do much better than `set_x`. Because, again, what does `shrink` and `fit` mean in this context. On the other hand if you passed in a rect that was already small enough to fit and wasn't the whole terminal, this would make no sense as you wouldn't need to shrink it. I need to think more about this. At a minimum, I need to improve the comments.

Reply:
Ok, I think I found a solution to our naming woes. I changed the behavior of the latter method(s) in 232e3ed8789e4a62369fe853318afd61267bec7b.

Instead of moving to an absolute position, `self` is explicitly treated as a bounding box and the new rect is shifted downwards by `y` columns. To stay within bounds, the height is shrunk by `y` columns. The new name of the method is `move_down_within_bounds` and it has a much better description. I also adopted your descriptions for the prior center methods and renamed them.

Reply:
Nice, I like that much more! :+1: